File: Trabalho3/cliente-python/cliente.py
import requests
import json

class SupermercadoCliente:
    def __init__(self, server_ip="localhost"):
        self.base_url = f"http://{server_ip}:8080"
        print(f"🔗 Conectando ao servidor: {self.base_url}")
    
    def testar_conexao(self):
        """Testar se consegue conectar ao servidor"""
        try:
            response = requests.get(f"{self.base_url}/", timeout=5)
            return response.status_code == 200
        except:
            return False
    
    def fazer_requisicao(self, metodo, endpoint, dados=None):
        """Fazer requisição HTTP"""
        url = f"{self.base_url}{endpoint}"
        
        try:
            if metodo == "GET":
                response = requests.get(url, timeout=10)
            elif metodo == "POST":
                headers = {"Content-Type": "application/json"}
                response = requests.post(url, json=dados, headers=headers, timeout=10)
            
            response.raise_for_status()
            return response.json()
        except requests.exceptions.ConnectionError:
            print(f"❌ Não foi possível conectar ao servidor: {self.base_url}")
            print("   Verifique:")
            print("   1. O servidor está rodando?")
            print("   2. O IP está correto?")
            print("   3. Ambos estão na mesma rede?")
            return None
        except Exception as e:
            print(f"❌ Erro: {e}")
            return None

def mostrar_menu():
    print("\n" + "="*50)
    print("🐍 CLIENTE PYTHON - SUPERMERCADO API")
    print("="*50)
    print("1. 📋 Listar todos os funcionários")
    print("2. 💰 Calcular folha de pagamento")
    print("3. 📊 Listar tipos de funcionários")
    print("4. 🔍 Filtrar por tipo")
    print("5. ➕ Cadastrar novo funcionário")
    print("6. 🌐 Testar conexão")
    print("7. 🚪 Sair")
    print("="*50)

def main():
    print("=== CONFIGURAÇÃO DE REDE ===")
    server_ip = input("Digite o IP do servidor (ou Enter para localhost): ").strip()
    if not server_ip:
        server_ip = "localhost"
    
    cliente = SupermercadoCliente(server_ip)
    
    # Testar conexão
    print(f"\n🔍 Testando conexão com {server_ip}...")
    if not cliente.testar_conexao():
        print("❌ Não foi possível conectar ao servidor!")
        print("   Verifique o IP e se o servidor está rodando.")
        return
    
    print("✅ Conexão estabelecida com sucesso!")
    
    while True:
        mostrar_menu()
        opcao = input("\nEscolha uma opção: ").strip()
        
        if opcao == "7":
            print("\n👋 Encerrando cliente...")
            break
        elif opcao == "6":
            if cliente.testar_conexao():
                print("✅ Servidor está respondendo!")
            else:
                print("❌ Servidor não está respondendo!")
            input("\nPressione Enter para continuar...")
            continue
        
        resultado = None
        
        if opcao == "1":
            print("\n📋 Listando funcionários...")
            resultado = cliente.fazer_requisicao("GET", "/api/funcionarios")
            if resultado:
                print(f"\n✅ Total: {len(resultado)} funcionários")
                for func in resultado:
                    print(f"\n  ID: {func.get('id')}")
                    print(f"  Nome: {func.get('nome')}")
                    print(f"  Tipo: {func.get('tipo')}")
                    print(f"  Salário: R$ {func.get('salario'):.2f}")
        
        elif opcao == "2":
            print("\n💰 Calculando folha...")
            resultado = cliente.fazer_requisicao("GET", "/api/funcionarios/folha")
            if resultado:
                print(f"\n✅ Folha total: R$ {resultado.get('folha_pagamento'):.2f}")
                print(f"✅ Funcionários: {resultado.get('quantidade_funcionarios')}")
                print(f"✅ Média salarial: R$ {resultado.get('media_salarios'):.2f}")
        
        elif opcao == "3":
            print("\n📊 Listando tipos...")
            resultado = cliente.fazer_requisicao("GET", "/api/funcionarios/tipos")
            if resultado:
                print("\n✅ Tipos disponíveis:")
                for tipo in resultado.get('tipos_disponiveis', []):
                    print(f"  • {tipo}")
        
        elif opcao == "4":
            tipo = input("\nDigite o tipo (Vendedor/Gerente/Caixa/Balconista): ").strip()
            print(f"\n🔍 Filtrando por {tipo}...")
            resultado = cliente.fazer_requisicao("GET", f"/api/funcionarios/tipo/{tipo}")
            if resultado:
                print(f"\n✅ Tipo: {resultado.get('tipo')}")
                print(f"✅ Quantidade: {resultado.get('quantidade')}")
                for func in resultado.get('funcionarios', []):
                    print(f"\n  ID: {func.get('id')}")
                    print(f"  Nome: {func.get('nome')}")
                    print(f"  Salário: R$ {func.get('salario'):.2f}")
        
        elif opcao == "5":
            print("\n➕ CADASTRAR NOVO FUNCIONÁRIO")
            print("="*30)
            
            dados = {}
            dados["nome"] = input("Nome: ").strip()
            
            while True:
                try:
                    salario = input("Salário (R$): ").replace(',', '.')
                    dados["salario"] = float(salario)
                    break
                except ValueError:
                    print("❌ Digite um número válido")
            
            tipos_validos = ["Vendedor", "Gerente", "Caixa", "Balconista"]
            while True:
                tipo = input("Tipo (Vendedor/Gerente/Caixa/Balconista): ").strip()
                if tipo in tipos_validos:
                    dados["tipo"] = tipo
                    break
                else:
                    print(f"❌ Tipo inválido. Escolha entre: {', '.join(tipos_validos)}")
            
            tipo = dados["tipo"]
            if tipo == "Vendedor":
                while True:
                    try:
                        comissao = input("Comissão (R$): ").replace(',', '.')
                        dados["comissao"] = float(comissao)
                        break
                    except ValueError:
                        print("❌ Digite um número válido")
            elif tipo == "Gerente":
                while True:
                    try:
                        bonus = input("Bônus (R$): ").replace(',', '.')
                        dados["bonus"] = float(bonus)
                        break
                    except ValueError:
                        print("❌ Digite um número válido")
            elif tipo == "Caixa":
                while True:
                    try:
                        dados["numeroCaixa"] = int(input("Número do caixa: "))
                        break
                    except ValueError:
                        print("❌ Digite um número inteiro")
            elif tipo == "Balconista":
                dados["setor"] = input("Setor: ").strip()
            
            print(f"\n📦 Enviando dados...")
            resultado = cliente.fazer_requisicao("POST", "/api/funcionarios", dados)
            
            if resultado:
                if resultado.get('success'):
                    print(f"\n🎉 {resultado.get('message')}")
                    print(f"   🆔 ID: {resultado.get('id')}")
                    print(f"   👤 Nome: {resultado.get('nome')}")
                    print(f"   📋 Tipo: {resultado.get('tipo')}")
                else:
                    print(f"\n❌ {resultado.get('message')}")
        
        else:
            print("\n❌ Opção inválida!")
        
        if opcao not in ["6", "7"]:
            input("\n⏎ Pressione Enter para continuar...")

if __name__ == "__main__":
    main()

File: Trabalho3/cliente-cpp/cliente.cpp
#include <iostream>
#include <curl/curl.h>
#include <string>
#include <sstream>
#include <cstring>

using namespace std;

size_t WriteCallback(void* contents, size_t size, size_t nmemb, string* output) {
    size_t totalSize = size * nmemb;
    output->append((char*)contents, totalSize);
    return totalSize;
}

bool testarConexao(const string& server_url) {
    CURL* curl = curl_easy_init();
    if(!curl) return false;
    
    string url = server_url + "/api/funcionarios";
    string response;
    
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);
    
    CURLcode res = curl_easy_perform(curl);
    curl_easy_cleanup(curl);
    
    return res == CURLE_OK;
}

int main() {
    CURL* curl;
    CURLcode res;
    string response;
    
    curl_global_init(CURL_GLOBAL_DEFAULT);
    curl = curl_easy_init();
    
    if(curl) {
        cout << "======================================" << endl;
        cout << "CLIENTE C++ - SUPERMERCADO API" << endl;
        cout << "======================================" << endl;
        
        // Configuração de rede
        cout << "\n=== CONFIGURAÇÃO DE REDE ===" << endl;
        cout << "Digite o IP do servidor (ou Enter para localhost): ";
        
        string server_ip;
        getline(cin, server_ip);
        if(server_ip.empty()) {
            server_ip = "localhost";
        }
        
        string server_url = "http://" + server_ip + ":8080";
        
        cout << "\n🔍 Testando conexão com " << server_url << "..." << endl;
        
        if(!testarConexao(server_url)) {
            cout << "❌ Não foi possível conectar ao servidor!" << endl;
            cout << "   Verifique:" << endl;
            cout << "   1. O servidor está rodando?" << endl;
            cout << "   2. O IP está correto? (Seu IP: 172.25.183.184)" << endl;
            cout << "   3. Ambos estão na mesma rede?" << endl;
            cout << "   4. Firewall pode estar bloqueando" << endl;
            
            curl_easy_cleanup(curl);
            curl_global_cleanup();
            return 1;
        }
        
        cout << "✅ Conexão estabelecida com sucesso!" << endl;
        
        while(true) {
            cout << "\n📋 MENU PRINCIPAL:" << endl;
            cout << "1. Listar todos os funcionários" << endl;
            cout << "2. Calcular folha de pagamento" << endl;
            cout << "3. Listar tipos de funcionários" << endl;
            cout << "4. Filtrar por tipo" << endl;
            cout << "5. Cadastrar novo funcionário" << endl;
            cout << "6. Testar conexão" << endl;
            cout << "7. Sair" << endl;
            cout << "\nEscolha uma opção: ";
            
            int opcao;
            cin >> opcao;
            cin.ignore();
            
            if(opcao == 7) {
                cout << "\n👋 Encerrando cliente..." << endl;
                break;
            }
            
            if(opcao == 6) {
                if(testarConexao(server_url))
                    cout << "\n✅ Servidor está respondendo!" << endl;
                else
                    cout << "\n❌ Servidor não está respondendo!" << endl;
                cout << "\nPressione Enter para continuar...";
                cin.get();
                continue;
            }
            
            string url;
            string post_data;
            
            switch(opcao) {
                case 1:
                    url = server_url + "/api/funcionarios";
                    cout << "\n📋 Listando todos os funcionários..." << endl;
                    break;
                    
                case 2:
                    url = server_url + "/api/funcionarios/folha";
                    cout << "\n💰 Calculando folha de pagamento..." << endl;
                    break;
                    
                case 3:
                    url = server_url + "/api/funcionarios/tipos";
                    cout << "\n📊 Listando tipos de funcionários..." << endl;
                    break;
                    
                case 4:
                    {
                        cout << "\n🔍 Tipos disponíveis: Vendedor, Gerente, Caixa, Balconista" << endl;
                        cout << "Digite o tipo: ";
                        string tipo;
                        getline(cin, tipo);
                        url = server_url + "/api/funcionarios/tipo/" + tipo;
                        cout << "\n🔍 Filtrando por tipo: " << tipo << endl;
                    }
                    break;
                    
                case 5:
                    {
                        cout << "\n➕ CADASTRAR NOVO FUNCIONÁRIO" << endl;
                        cout << "=============================" << endl;
                        
                        string nome, tipo;
                        double salario;
                        
                        cout << "Nome: ";
                        getline(cin, nome);
                        
                        cout << "Salário (R$): ";
                        cin >> salario;
                        cin.ignore();
                        
                        cout << "Tipo (Vendedor/Gerente/Caixa/Balconista): ";
                        getline(cin, tipo);
                        
                        // Construir JSON CORRETAMENTE
                        stringstream json_stream;
                        json_stream << "{";
                        json_stream << "\"nome\":\"" << nome << "\",";
                        json_stream << "\"salario\":" << salario << ",";
                        json_stream << "\"tipo\":\"" << tipo << "\"";
                        
                        // Adicionar campos específicos
                        if(tipo == "Vendedor") {
                            double comissao;
                            cout << "Comissão (R$): ";
                            cin >> comissao;
                            cin.ignore();
                            json_stream << ",\"comissao\":" << comissao;
                        }
                        else if(tipo == "Gerente") {
                            double bonus;
                            cout << "Bônus (R$): ";
                            cin >> bonus;
                            cin.ignore();
                            json_stream << ",\"bonus\":" << bonus;
                        }
                        else if(tipo == "Caixa") {
                            int numCaixa;
                            cout << "Número do caixa: ";
                            cin >> numCaixa;
                            cin.ignore();
                            json_stream << ",\"numeroCaixa\":" << numCaixa;
                        }
                        else if(tipo == "Balconista") {
                            string setor;
                            cout << "Setor: ";
                            getline(cin, setor);
                            json_stream << ",\"setor\":\"" << setor << "\"";
                        }
                        
                        json_stream << "}";
                        post_data = json_stream.str();
                        url = server_url + "/api/funcionarios";
                        
                        cout << "\n📦 JSON enviado: " << post_data << endl;
                    }
                    break;
                    
                default:
                    cout << "\n❌ Opção inválida!" << endl;
                    continue;
            }
            
            // Configurar requisição
            curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
            curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
            curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
            
            // Se for POST, configurar dados
            if(opcao == 5) {
                curl_easy_setopt(curl, CURLOPT_POST, 1L);
                curl_easy_setopt(curl, CURLOPT_POSTFIELDS, post_data.c_str());
                
                struct curl_slist* headers = NULL;
                headers = curl_slist_append(headers, "Content-Type: application/json");
                curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
            }
            
            // Executar requisição
            res = curl_easy_perform(curl);
            
            if(res != CURLE_OK) {
                cerr << "\n❌ Erro na requisição: " << curl_easy_strerror(res) << endl;
            } else {
                cout << "\n✅ Resposta do servidor:" << endl;
                cout << "========================" << endl;
                cout << response << endl;
                cout << "========================" << endl;
            }
            
            // Limpar resposta
            response.clear();
            
            if(opcao == 5) {
                // Limpar headers do POST
                struct curl_slist* headers = NULL;
                curl_slist_free_all(headers);
            }
            
            cout << "\nPressione Enter para continuar...";
            cin.get();
        }
        
        // Limpar
        curl_easy_cleanup(curl);
    }
    
    curl_global_cleanup();
    return 0;
}